src/manifest.js

```diff
@@ -0,0 +1,157 @@
+import fs from 'fs';
+import * as utils from './utils/index.js';
+import zlib from 'zlib';
+import { promisify } from 'util';
+import AWS from 'aws-sdk';
+
+const gunzip = promisify(zlib.gunzip)
+
+const LATEST = "LATEST"
+
+const {updateResourceVersions, ResourceVersions} = (() => {
+
+  const data = {};
+  const eTag = {value: undefined};
+
+  /**
+   * We expose an update function rather than automatically updating here. This
+   * prevents scripts / tests which (indirectly) import this file unintentionally
+   * fetching stuff from S3 (and often failing due to missing credentials).
+   */
+  const updateResourceVersions = () => _update(data, eTag);
+
+  class ResourceVersions {
+    constructor(resourceType, resourceId) {
+      this.data = {};
+      if (!data[resourceType]) {
+        utils.verbose(`Attempted to access resources of type ${resourceType} but there are none in the manifest`);
+        return;
+      }
+      this.data = data[resourceType][resourceId] || {};
+    }
+
+    /**
+     * Given a YYYY-MM-DD string return one of four values:
+     * - LATEST: if the date is more recent that any in the manifest
+     * - false: there are no records in the manifest, or the dateDescriptor
+     *          is mis-formed
+     * - YYYY-MM-DD string: the closest match (may be an exact match)
+     * 
+     * In the future this will be expanded to parse other kinds of
+     * dateDescriptors (e.g. a descriptor meaning "6 months ago").
+     * 
+     * When there is not an exact match we interpret closest to be
+     * the closest but more recent version.
+     */
+    closestDate(dateDescriptor) {
+      if (!dateDescriptor.match(/^\d{4}-\d{2}-\d{2}$/)) {
+        return false;
+      }
+      const availableDates = (this.data?.versions || []).map((v) => v.date);
+      if (!availableDates.length) {
+        return false;
+      }
+      let closest = availableDates[0];
+      if (closest < dateDescriptor) {
+        return LATEST;
+      }
+      for (const date of availableDates) {
+        if (date >= dateDescriptor) {
+          closest = date;
+        } else {
+          break;
+        }
+      }
+      return closest;
```

Comment:
> This algorithm depends on the order of .versions in the resources index. It should explicitly sort its input if it's going to rely on a specific order.

Yup! The versions are explicitly sorted by date during index creation (e.g. see `createVersionedResources`)

Reply:
Yep, I know they're sorted by date during index creation, but that's distant from this code and it would be an easy property of the index to accidentally break for a variety of reasons. I'm suggesting to sort here by date explicitly, regardless of what happens during index creation, to ensure the required sorting invariant. If sorting is demonstrably too expensive to do here on each request, then it would be ok to me to do it just once in the less-distant code which loads the index into memory. But better to assert that invariant here, as close as possible to the requirement, if the overhead is not too much. (I expect it's fine, esp. if the sorting algorithm (hopefully) has better performance on already sorted input.)

Reply:
Ok. I'm not worried about performance - many of our server responses involve waiting on their own network calls which will be orders of magnitude slower. Done.